src/audio/host-zephyr.c

```diff
@@ -676,44 +591,80 @@ static struct comp_dev *host_new(const struct comp_driver *drv,
 	if (!hd->dma) {
 		comp_err(dev, "host_new(): dma_get() returned NULL");
 		rfree(hd);
-		rfree(dev);
-		return NULL;
+		return -ENODEV;
 	}
 
 	/* init buffer elems */
 	dma_sg_init(&hd->config.elem_array);
 	dma_sg_init(&hd->host.elem_array);
 	dma_sg_init(&hd->local.elem_array);
 
-	ipc_build_stream_posn(&hd->posn, SOF_IPC_STREAM_POSITION, dev->ipc_config.id);
+	ipc_build_stream_posn(&hd->posn, SOF_IPC_STREAM_POSITION, config_id);
 
 	hd->msg = ipc_msg_init(hd->posn.rhdr.hdr.cmd, sizeof(hd->posn));
 	if (!hd->msg) {
 		comp_err(dev, "host_new(): ipc_msg_init failed");
 		dma_put(hd->dma);
 		rfree(hd);
+		return -ENOMEM;
+	}
+	hd->chan = NULL;
+	hd->copy_type = COMP_COPY_NORMAL;
+
+	return 0;
+}
+
+static struct comp_dev *host_new(const struct comp_driver *drv,
+				 const struct comp_ipc_config *config,
+				 const void *spec)
+{
+	struct comp_dev *dev;
+	struct host_data *hd;
+	const struct ipc_config_host *ipc_host = spec;
+	int ret;
+
+	comp_cl_dbg(&comp_host, "host_new()");
+
+	dev = comp_alloc(drv, sizeof(*dev));
+	if (!dev)
+		return NULL;
+	dev->ipc_config = *config;
+
+	hd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*hd));
+	if (!hd) {
 		rfree(dev);
 		return NULL;
 	}
+	comp_set_drvdata(dev, hd);
+
+	ret = host_zephyr_new(hd, dev, ipc_host, dev->ipc_config.id);
+	if (ret) {
+		rfree(hd);
+		rfree(dev);
+		comp_err(dev, "host: host zephyr new dma failed with exit");
```

Comment:
what does "new dma failed" mean? I dont think you need an error here. host_zephyr_new() prints an error message in all its error paths already

Reply:
ok, removing